Justin projects performance test

SWYZJustin opened the pull request:
Just to make sure that this performance test is in the right format. The commented method sometimes don't work correctly when running the tests togather. I will change it later.

SWYZJustin committed:
Create ProjectsPerformanceTest.java

SWYZJustin committed:
Performance Test for Projects

I push this just to make sure that it is in the right format.

Lobo808 commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,342 @@
+            timeTable.add(time);
+        }
+        System.out.println(timeTable);
+        Runtime runtime = Runtime.getRuntime();
+        runtime.gc();
```
Not sure what this does... I used debug mode and it seems like runtime gives zero information

SWYZJustin replied:
I don't know actually. I am just looking for ways to maks the comment part work. I will modify it later. Thank u for your comments.

Lobo808 commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,342 @@
+            Response response = RestAssured.given().contentType(ContentType.JSON).body(body.toString()).post(pUrl);
+            long end = System.nanoTime();
+            long time = end-start;
+            timeTable.add(time);
```
I think saving 1000 values is a bit too heavy and difficult to see. Maybe only record time when the server has 1, 10, 50, 100, 200, 500, 1000 projects (or something like that). We can better visualize the trend from there.

SWYZJustin replied:
Thank you for your suggestion. I will change it later.

Lobo808 reviewed (changes requested):
Only left comments for the first test, but it applies for everything in general

Aryan-Akhilesh commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,342 @@
+        runtime.gc();
+    }
+
+    @Test
+    public void createProjectsJsonPut(){
+        List<Long> timeTable = new ArrayList<Long>();
+        for(int i=0;i<1000;i++){
+            JSONObject body = new JSONObject();
+            String newTitle = "New project " + i;
+            Boolean newCompleted = false;
+            Boolean newActive = false;
+            String newDescription = "New Description";
+            body.put("title",newTitle);
+            body.put("completed",newCompleted);
+            body.put("active",newActive);
+            body.put("description",newDescription);
+            long start = System.nanoTime();
+            Response response = RestAssured.given().contentType(ContentType.JSON).body(body.toString()).put(pUrl);
+            long end = System.nanoTime();
+            long time = end-start;
+            timeTable.add(time);
+        }
+        System.out.println(timeTable);
+        Runtime runtime = Runtime.getRuntime();
+        runtime.gc();
+    }
```
This doesn't create data. PUT is not allowed with the API endpoint you are using. The response has a status code of 405

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,342 @@
+            Response response = RestAssured.given().contentType(ContentType.XML).body(xmlBody).put(pUrl);
+            long end = System.nanoTime();
+            long time = end-start;
+            timeTable.add(time);
+        }
+        System.out.println(timeTable);
```
if you are adding to a table and then printing all at the end. Wouldn't it be more optimal to print it each time when needed (e.g i == 0, i == 1000...). This way you won't need to have a list

SWYZJustin committed:
Update Performance Test

Update performance test based on comments

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,203 @@
+    private static ProcessBuilder pb;
+
+    private final String json = "application/json";
+    private final String xml = "application/xml";
```
if you didn't use it remove it

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,203 @@
+
+
+    @Test
+    public void createProjectsJsonPut() {
```
pretty sure that for creating you need Post

SWYZJustin replied:
Sure. I will change it.

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Test
+    public void amendProjectsJsonPost(){
```
For amend you need put

SWYZJustin replied:
Thanks. I will change it.

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,203 @@
+            long time = end-start;
+            double time_in_second = (double) time / 1_000_000_000;
+            double cpuUsage = osBean.getCpuLoad() * 100;
+            long memory = osBean.getFreeMemorySize()/ (1024L * 1024L);
```
logic is wrong here, you are populating everthing and then you modify, but the correct step should be amending when populating.

SWYZJustin replied:
Let me have a look.

MosinLover commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,203 @@
+                freeMemoryStore[index] = memory;
+                index++;
+            }
+        }
```
logic is wrong here since you populated and then delete one by one thus your order at final will be wrong, you might want to check how jialin did with his second for loop.

SWYZJustin replied:
Alright. I will check it.

SWYZJustin committed:
Update according to the comment

Change the logic, delete unnecessary lines, change the create with post and amend with put, add some comments

Lobo808 commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,211 @@
+        for(int i=0;i<tot;i++){
+            JSONObject body = new JSONObject();
+            String newTitle = "New project " + i;
+            Boolean newCompleted = false;
+            Boolean newActive = false;
+            String newDescription = "New Description";
+            body.put("title",newTitle);
+            body.put("completed",newCompleted);
+            body.put("active",newActive);
+            body.put("description",newDescription);
+            Response r1 = RestAssured.given().contentType(ContentType.JSON).body(body.toString()).post(pUrl);
+            String id = r1.jsonPath().get("id");
+
+
+            long start = System.nanoTime();
+            Response r2 = RestAssured.given().delete(pUrl+"/"+id);
+            long end = System.nanoTime();
+            long time = end-start;
+            double time_in_second = (double) time / 1_000_000_000;
+            double cpuUsage = osBean.getCpuLoad() * 100;
+            long memory = osBean.getFreeMemorySize()/ (1024L * 1024L);
+            if(i == target[index]-1){
+                timeStore[index] = time_in_second;
+                cpuUsageStore[index] = cpuUsage;
+                freeMemoryStore[index] = memory;
+                index++;
+            }
+        }
+        System.out.println("---------Delete Project Statistics---------");
+        System.out.printf("%-10s %-20s %-20s %-20s%n", "SIZE", "TIME (s)", "CPU USAGE (%)", "MEMORY (MB)");
+        for (int i = 0; i < target.length; i++) {
+            System.out.printf("%-10d %-20f %-20f %-20d%n",
+                    target[i],
+                    timeStore[i],
+                    cpuUsageStore[i],
+                    freeMemoryStore[i]);
+        }
+        System.out.println("----------------------------------------");
+    }
+}
```
You are deleting the object upon creating. How is this testing different numbers of objects?

SWYZJustin replied:
Sure. I will change it.

Lobo808 reviewed (changes requested):
you still have quite a few unused variables. I didn't mark them in the code, but please delete them.

SWYZJustin commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,342 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+@TestMethodOrder(MethodOrderer.Random.class)
```
Thanks.

SWYZJustin committed:
Change back the logic for 'Delete' test

Lobo808 commented on TodoListManagerTest/src/test/java/PerformanceTest/ProjectsPerformanceTest.java:
```diff
@@ -0,0 +1,211 @@
+            String id = r1.jsonPath().get("id");
+            idList[i] = id;
+        }
+        for(int i=0;i<tot;i++){
+            long start = System.nanoTime();
+            RestAssured.given().delete(pUrl+"/"+ idList[tot-1-i]);
+            long end = System.nanoTime();
+            long time = end-start;
+            double time_in_second = (double) time / 1_000_000_000;
+            double cpuUsage = osBean.getCpuLoad() * 100;
+            long memory = osBean.getFreeMemorySize()/ (1024L * 1024L);
+            if(i == target[index]-1){
+                timeStore[index] = time_in_second;
+                cpuUsageStore[index] = cpuUsage;
+                freeMemoryStore[index] = memory;
+                index++;
+            }
+        }
```
still wrong... you are storing the time for 1000 in the slot for 1 (basically everything reversed)

SWYZJustin committed:
Change the delete logic

Change the delete logic. This time it should be correct.

Lobo808 committed:
remove 5 from target size

Lobo808 commented:
lgtm
